src/compiler/scala/tools/nsc/transform/patmat/MatchTranslation.scala

```diff
@@ -110,20 +110,28 @@ trait MatchTranslation {
         val (makers, unappBinder) = {
           val paramType = extractor.expectedExtractedType
           // Statically conforms to paramType
-          if (tpe <:< paramType) (treeMakers(binder, false, pos), binder)
+          if (tpe <:< paramType) {
+            // enforce all extractor patterns to be non-null
+            val nonNullTest = NonNullTestTreeMaker(binder, paramType, pos)
+            val unappBinder = nonNullTest.nextBinder
+            (nonNullTest :: treeMakers(unappBinder, pos), unappBinder)
+          }
           else {
             // chain a type-testing extractor before the actual extractor call
             // it tests the type, checks the outer pointer and casts to the expected type
             // TODO: the outer check is mandated by the spec for case classes, but we do it for user-defined unapplies as well [SPEC]
             // (the prefix of the argument passed to the unapply must equal the prefix of the type of the binder)
             val typeTest = TypeTestTreeMaker(binder, binder, paramType, paramType)(pos, extractorArgTypeTest = true)
             val binderKnownNonNull = typeTest impliesBinderNonNull binder
-
-            // check whether typetest implies binder is not null,
-            // even though the eventual null check will be on typeTest.nextBinder
-            // it'll be equal to binder casted to paramType anyway (and the type test is on binder)
-            val unappBinder = typeTest.nextBinder
-            (typeTest :: treeMakers(unappBinder, binderKnownNonNull, pos), unappBinder)
+            // skip null test if it's implied
+            if (binderKnownNonNull) {
+              val unappBinder = typeTest.nextBinder
+              (typeTest :: treeMakers(unappBinder, pos), unappBinder)
+            } else {
+              val nonNullTest = NonNullTestTreeMaker(typeTest.nextBinder, paramType, pos)
+              val unappBinder = nonNullTest.nextBinder
+              (typeTest :: nonNullTest :: treeMakers(unappBinder, pos), unappBinder)
```

Comment:
@adriaanm Instead of doing null check in `ExtractorTreeMaker` (like I had in b6a007d434cdeb95321be16b8d3274a17318cf8c), I've created a new tree maker `NonNullTestTreeMaker`. This is  positioned _after_ `typeTest`. If `binderKnowNonNull` is `true`, I am skipping the test.

I've removed null checking in `ProductExtractorTreeMaker` and passing `binderKnowNonNull` around since it's now redundant.